examples/water_sampling_hrex.py

```diff
@@ -0,0 +1,355 @@
+
+    # read water system
+    solvent_sys, solvent_conf, solvent_box, solvent_topology, num_water_atoms = build_system(
+        water_pdb, ff.water_ff, padding=0.1
+    )
+    solvent_box += np.diag([0.1, 0.1, 0.1])  # remove any possible clashes
+
+    assert num_water_atoms == len(solvent_conf)
+
+    bt = BaseTopology(mol, ff)
+    afe = AbsoluteFreeEnergy(mol, bt)
+    host_config = HostConfig(solvent_sys, solvent_conf, solvent_box, num_water_atoms)
+    potentials, params, combined_masses = afe.prepare_host_edge(ff.get_params(), host_config, lamb)
+
+    ligand_idxs = np.arange(num_water_atoms, num_water_atoms + mol.GetNumAtoms())
+
+    # re-parameterize ligand idxs to do decouple only half of the buckyball
+    nb_params = params[-1]
+
+    final_masses = combined_masses
+    final_conf = np.concatenate([solvent_conf, get_romol_conf(mol)], axis=0)
+
+    component_dim = 4  # q,s,e,w
+    num_atoms = len(final_conf)
+
+    ligand_water_flat_idxs = np.s_[component_dim * num_atoms : 2 * component_dim * num_atoms]  # water
+    ligand_protein_flat_idxs = np.s_[2 * component_dim * num_atoms : 3 * component_dim * num_atoms :]  # non-water
+    ligand_water_params = nb_params[ligand_water_flat_idxs].reshape(-1, 4)
+    ligand_protein_params = nb_params[ligand_protein_flat_idxs].reshape(-1, 4)
+
+    fully_coupled_ligand_atoms = ligand_idxs[: len(ligand_idxs) // 2]
+
+    ligand_water_params = ligand_water_params.at[fully_coupled_ligand_atoms, -1].set(0)
+    ligand_protein_params = ligand_protein_params.at[fully_coupled_ligand_atoms, -1].set(0)
+
+    nb_params = nb_params.at[ligand_water_flat_idxs].set(ligand_water_params.reshape(-1))
+    nb_params = nb_params.at[ligand_protein_flat_idxs].set(ligand_protein_params.reshape(-1))
+
+    # override last potential with the updated nb_params
+    host_bps = []
+    for p, bp in zip(params[:-1], potentials[:-1]):
+        host_bps.append(bp.bind(p))
+
+    host_bps.append(potentials[-1].bind(nb_params))
+
+    # (YTZ): This is disabled because the initial ligand and starting waters are pre-minimized
+    # print("Minimizing the host system...", end="", flush=True)
+    # host_conf = minimize_host_4d([mol], host_config, ff)
+    # print("Done", flush=True)
+
+    temperature = DEFAULT_TEMP
+    dt = 1e-3
+    barostat_interval = 25
+    integrator = LangevinIntegrator(temperature, dt, 1.0, final_masses, seed)
+    bond_list = get_bond_list(host_bps[0].potential)
+    group_idxs = get_group_indices(bond_list, len(final_masses))
+    barostat = MonteCarloBarostat(
+        len(final_masses), DEFAULT_PRESSURE, temperature, group_idxs, barostat_interval, seed + 1
+    )
+
+    initial_state = InitialState(
+        host_bps, integrator, barostat, final_conf, np.zeros_like(final_conf), solvent_box, lamb, ligand_idxs
+    )
+
+    num_water_mols = num_water_atoms // 3
+
+    return initial_state, num_water_mols, solvent_topology
+
+
+def compute_density(n_waters, box):
```

Comment:
Would be good to specify the units in a docstring

Reply:
willfix